server/container_create_linux.go

```diff
@@ -228,6 +228,15 @@ func (s *Server) createSandboxContainer(ctx context.Context, ctr ctrfactory.Cont
 		someRepoDigest = imgResult.RepoDigests[0]
 	}
 
+	systemCtx, err := s.contextForNamespace(sb.Metadata().Namespace)
+	if err != nil {
+		return nil, fmt.Errorf("get context for namespace: %w", err)
+	}
+
+	if err := s.StorageImageServer().IsRunningImageAllowed(ctx, &systemCtx, *imageName, imgResult.Digest); err != nil {
```

Comment:
⚠️ 

Not a review, a very minimal look: This is incorrect. The signature must be verified against users’ intent (I think that’s `ctr.UserRequestedImage` but I didn’t now verify that). This does it against `SomeNameOfThisImage`, which is _explicitly documented_ to not be usable.

I … thought I might have gone overboard with the verbose `SomeNameOf` naming or with the shouting in the doc strings, it clearly isn’t sufficient. Can you suggest some other way to make these warnings stick?

Reply:
And `SomeNameOfThisImage` is, also, explicitly documented to not always be available, as suggested by the `*string` type.

Reply:
```golang
	userRequestedImage, err := ctr.UserRequestedImage()
	if err != nil {
		return nil, err
	}
	// Get imageName and imageID that are later requested in container status
	var imgResult *storage.ImageResult
	if id := s.StorageImageServer().HeuristicallyTryResolvingStringAsIDPrefix(userRequestedImage); id != nil {
		imgResult, err = s.StorageImageServer().ImageStatusByID(s.config.SystemContext, *id)
		if err != nil {
			return nil, err
		}
	} else {
		potentialMatches, err := s.StorageImageServer().CandidatesForPotentiallyShortImageName(s.config.SystemContext, userRequestedImage)
		if err != nil {
			return nil, err
		}
		var imgResultErr error
		for _, name := range potentialMatches {
			imgResult, imgResultErr = s.StorageImageServer().ImageStatusByName(s.config.SystemContext, name)
			if imgResultErr == nil {
				break
			}
		}
		if imgResultErr != nil {
			return nil, imgResultErr
		}
	}
	// At this point we know userRequestedImage is not empty; "" is accepted by neither HeuristicallyTryResolvingStringAsIDPrefix
	// nor CandidatesForPotentiallyShortImageName. Just to be sure:
	if userRequestedImage == "" {
		return nil, errors.New("internal error: successfully found an image, but userRequestedImage is empty")
	}

	imageName := imgResult.SomeNameOfThisImage
```

`imageName` is obtained from `imgResult` which in turn comes from `ctr.UserRequestedImage()`

Reply:
But let us see how to get `imageName` from `imageResult` without using `SomeNameOfThisImage`. Any pointer is highly appreciated.

Reply:
`references.ParseRegistryImageReferenceFromOutOfProcessData(userRequestedImage)` might work, let me test it.

Reply:
The name to enforce policy against is an _input_ to the security decision whether to trust an image. It logically does not _depend_ on image lookup. Looking for it in `imageResult` is structurally incorrect already.

---

Even without that conceptual thought, just looking at the shape of the code:

- In the `ImageStatusByID` case, the input structurally isn’t a name at all, so there is no way for it to end up in `SomeNameOfThisImage` in a way that could work.
- If you read how `ImageStatusByName` (or `…ByID`, it’s the same) constructs `imgResult`, the input `name` parameter does not participate in the construction of `SomeNameOfThisImage` at all.

---

Actually, `UserRequestedImage` is not the right field, it seems (it’s the output of Kubelet’s ImagePull and/or ImageStatus, so it can contain a non-name ID, as in the `ImageStatusByID` case); the policy needs to be driven by `UserSpecifiedImage`,  added by https://github.com/kubernetes/kubernetes/pull/118652 extra for this purpose. ( https://github.com/cri-o/cri-o/pull/7046/files does use that field .)

Reply:
Thanks, I changed to code to use `userSpecifiedImage := ctr.Config().GetImage().UserSpecifiedImage`  instead of  `UserRequestedImage`

Reply:
Also it uses `userSpecifiedImageReference, err := references.ParseRegistryImageReferenceFromOutOfProcessData(userSpecifiedImage)` to contruct `RegistryImageReference` instead of relying on `SomeNameOfThisImage`

Reply:
passing all policy bats tests,

```
ok 1 accept unsigned image with default policy
ok 2 deny unsigned image with restrictive policy
ok 3 deny unsigned image with restrictive policy if already pulled
ok 4 accept signed image with default policy
ok 5 accept signed image with restrictive policy
ok 6 accept unsigned image with not existing namespace policy
ok 7 accept unsigned image with higher priority namespace policy
ok 8 deny unsigned image with higher priority namespace policy
ok 9 accept signed image with higher priority namespace policy
ok 10 allow signed image with restrictive policy on container creation1 (fresh pull)
ok 11 deny unsigned image with restrictive policy on container creation2 (fresh pull)
ok 12 allow signed image with restrictive policy on container creation3 if already pulled (by ID)
ok 13 deny unsigned image with restrictive policy on container creation4 if already pulled (by ID)
ok 14 allow signed image with restrictive policy on container creation5 if already pulled (by tag)
ok 15 deny unsigned image with restrictive policy on container creation6 if already pulled (by tag)
```

Reply:
/hold till the CI passes.